relay-server/src/actors/project_cache.rs

```diff
@@ -400,6 +374,109 @@ struct UpdateProjectState {
     no_cache: bool,
 }
 
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Clone, Copy)]
+struct QueueKey {
+    key: ProjectKey,
+    sampling_key: ProjectKey,
+}
+
+/// The queue (buffer) of the incoming envelopes.
+#[derive(Debug, Default)]
+struct Queue {
+    /// Contains the cache of the incoming envelopes.
+    buffer: HashMap<QueueKey, Vec<(Box<Envelope>, EnvelopeContext)>>,
+    /// Index of the buffered project keys.
+    index: HashMap<ProjectKey, HashSet<QueueKey>>,
+}
+
+impl Queue {
+    /// Creates an empty queue.
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Adds the value to the queue for the provided key.
+    pub fn enqueue(&mut self, key: QueueKey, value: (Box<Envelope>, EnvelopeContext)) {
+        // Add key to our index.
+        for partial_key in &[key.key, key.sampling_key] {
+            self.index
+                .entry(*partial_key)
+                .or_insert(HashSet::from([key]))
+                .insert(key);
+        }
+
+        match self.buffer.entry(key) {
+            Entry::Occupied(o) => o.into_mut().push(value),
+            Entry::Vacant(v) => {
+                v.insert(vec![value]);
+            }
+        }
+    }
+
+    /// Returns the list of buffered envelopes if they satisfy a predicate.
+    pub fn dequeue<P>(
+        &mut self,
+        partial_key: &ProjectKey,
+        f: P,
+    ) -> Vec<(Box<Envelope>, EnvelopeContext)>
+    where
+        P: Fn(&QueueKey) -> bool,
+    {
+        let mut keys = self
+            .index
+            .get_mut(partial_key)
+            .map(mem::take)
```

Comment:
This leaves an empty set on the index, so you might as well remove it. Or does the empty set represent some special state?
```suggestion
            .remove(partial_key)
```